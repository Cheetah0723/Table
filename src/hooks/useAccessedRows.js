import { useMemo } from 'react'

export default function useAccessedRows ({
  debug, data, columns, subRowsKey,
}) {
  return useMemo(
    () => {
      if (debug) console.info('getAccessedRows')

      // Access the row's data
      const accessRow = (data, i, depth = 0) => {
        // Keep the original reference around
        const original = data

        // Process any subRows
        const subRows = data[subRowsKey]
          ? data[subRowsKey].map((d, i) => accessRow(d, i, depth + 1))
          : undefined

        const row = {
          original,
          index: i,
          subRows,
          depth,
        }

        // Create the cells and values
        row.values = {}
        columns.forEach(column => {
          row.values[column.id] = column.accessor ? column.accessor(data) : undefined
        })

        return row
      }

      // Use the resolved data
      return data.map((d, i) => accessRow(d, i))
    },
    [data, columns]
  )
}
