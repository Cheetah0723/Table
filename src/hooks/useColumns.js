import { useMemo } from 'react'
import PropTypes from 'prop-types'

import { getBy } from '../utils'

const propTypes = {
  // General
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      Cell: PropTypes.any,
      Header: PropTypes.any,
    })
  ),
  defaultColumn: PropTypes.any,
}

// Find the depth of the columns
function findMaxDepth(columns, depth = 0) {
  return columns.reduce((prev, curr) => {
    if (curr.columns) {
      return Math.max(prev, findMaxDepth(curr.columns, depth + 1))
    }
    return depth
  }, 0)
}

function decorateColumn(column, defaultColumn, parent, depth, index) {
  // Apply the defaultColumn
  column = { ...defaultColumn, ...column }

  // First check for string accessor
  let { id, accessor, Header } = column

  if (typeof accessor === 'string') {
    id = id || accessor
    const accessorString = accessor
    accessor = row => getBy(row, accessorString)
  }

  if (!id && typeof Header === 'string') {
    id = Header
  }

  if (!id && column.columns) {
    console.error(column)
    throw new Error('A column ID (or unique "Header" value) is required!')
  }

  if (!id) {
    console.error(column)
    throw new Error('A column ID (or string accessor) is required!')
  }

  column = {
    Header: ({ id }) => id,
    Cell: ({ value }) => value,
    show: true,
    ...column,
    id,
    accessor,
    parent,
    depth,
    index,
  }

  return column
}

// Build the visible columns, headers and flat column list
function decorateColumnTree(columns, defaultColumn, parent, depth = 0) {
  return columns.map((column, columnIndex) => {
    column = decorateColumn(column, defaultColumn, parent, depth, columnIndex)
    if (column.columns) {
      column.columns = decorateColumnTree(
        column.columns,
        defaultColumn,
        column,
        depth + 1
      )
    }
    return column
  })
}

// Build the header groups from the bottom up
function makeHeaderGroups(columns, maxDepth, defaultColumn) {
  const headerGroups = []

  const removeChildColumns = column => {
    delete column.columns
    if (column.parent) {
      removeChildColumns(column.parent)
    }
  }
  columns.forEach(removeChildColumns)

  const buildGroup = (columns, depth = 0) => {
    const headerGroup = {
      headers: [],
    }

    const parentColumns = []

    const hasParents = columns.some(col => col.parent)

    columns.forEach(column => {
      const isFirst = !parentColumns.length
      let latestParentColumn = [...parentColumns].reverse()[0]

      // If the column has a parent, add it if necessary
      if (column.parent) {
        if (isFirst || latestParentColumn.originalID !== column.parent.id) {
          parentColumns.push({
            ...column.parent,
            originalID: column.parent.id,
            id: [column.parent.id, parentColumns.length].join('_'),
          })
        }
      } else if (hasParents) {
        // If other columns have parents, add a place holder if necessary
        const placeholderColumn = decorateColumn(
          {
            originalID: [column.id, 'placeholder', maxDepth - depth].join('_'),
            id: [
              column.id,
              'placeholder',
              maxDepth - depth,
              parentColumns.length,
            ].join('_'),
          },
          defaultColumn
        )
        if (
          isFirst ||
          latestParentColumn.originalID !== placeholderColumn.originalID
        ) {
          parentColumns.push(placeholderColumn)
        }
      }

      // Establish the new columns[] relationship on the parent
      if (column.parent || hasParents) {
        latestParentColumn = [...parentColumns].reverse()[0]
        latestParentColumn.columns = latestParentColumn.columns || []
        if (!latestParentColumn.columns.includes(column)) {
          latestParentColumn.columns.push(column)
        }
      }

      headerGroup.headers.push(column)
    })

    headerGroups.push(headerGroup)

    if (parentColumns.length) {
      buildGroup(parentColumns)
    }
  }

  buildGroup(columns)

  return headerGroups.reverse()
}

export const useColumns = props => {
  const {
    debug,
    columns: userColumns,
    defaultColumn = {},
    state: [{ groupBy }],
  } = props

  PropTypes.checkPropTypes(propTypes, props, 'property', 'useColumns')

  const { columns, headerGroups, headers } = useMemo(() => {
    if (debug) console.info('getColumns')

    // Decorate All the columns
    let columnTree = decorateColumnTree(userColumns, defaultColumn)

    // Get the flat list of all columns
    let columns = flattenBy(columnTree, 'columns')

    columns = [
      ...groupBy.map(g => columns.find(col => col.id === g)),
      ...columns.filter(col => !groupBy.includes(col.id)),
    ]

    // Get headerGroups
    const headerGroups = makeHeaderGroups(
      columns,
      findMaxDepth(columnTree),
      defaultColumn
    )

    const headers = flattenBy(headerGroups, 'headers')

    return {
      columns,
      headerGroups,
      headers,
    }
  }, [debug, defaultColumn, groupBy, userColumns])

  return {
    ...props,
    columns,
    headerGroups,
    headers,
  }

  function flattenBy(columns, childKey) {
    const flatColumns = []

    const recurse = columns => {
      columns.forEach(d => {
        if (!d[childKey]) {
          flatColumns.push(d)
        } else {
          recurse(d[childKey])
        }
      })
    }

    recurse(columns)

    return flatColumns
  }
}
