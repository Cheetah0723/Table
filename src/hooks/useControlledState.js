const mergeStateAndProps = (state, defaults, props) => {
  Object.keys(props).forEach(key => {
    if (typeof props[key] !== 'undefined') {
      state[key] = props[key]
    }
  })

  Object.keys(defaults).forEach(key => {
    if (typeof state[key] === 'undefined') {
      state[key] = defaults[key]
    }
  })

  return state
}

export default function useControlledState ([userState, userSetState], defaults, props) {
  const state = mergeStateAndProps(userState, defaults, props)

  const setState = (updater, type) =>
    userSetState(old => updater(mergeStateAndProps(old, defaults, props)), type)

  return [state, setState]
}
