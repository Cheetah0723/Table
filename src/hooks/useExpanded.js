import { useMemo } from 'react'
import PropTypes from 'prop-types'

import { getBy, getFirstDefined, setBy } from '../utils'
import { addActions, actions } from '../actions'
import { defaultState } from './useTableState'

defaultState.expanded = {}

addActions({
  toggleExpanded: '__toggleExpanded__',
  useExpanded: '__useExpanded__'
})

const propTypes = {
  expandedKey: PropTypes.string,
  paginateSubRows: PropTypes.bool
}

export const useExpanded = props => {
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useExpanded')

  const {
    debug,
    rows,
    expandedKey = 'expanded',
    hooks,
    state: [{ expanded }, setState],
    paginateSubRows = true
  } = props

  const toggleExpandedByPath = (path, set) => {
    return setState(old => {
      const { expanded } = old
      const existing = getBy(expanded, path)
      set = getFirstDefined(set, !existing)
      return {
        ...old,
        expanded: setBy(expanded, path, set)
      }
    }, actions.toggleExpanded)
  }

  hooks.row.push(row => {
    const { path } = row
    row.toggleExpanded = set => toggleExpandedByPath(path, set)
  })

  const expandedRows = useMemo(() => {
    if (debug) console.info('getExpandedRows')

    const expandedRows = []

    // Here we do some mutation, but it's the last stage in the
    // immutable process so this is safe
    const handleRow = (row, index, depth = 0, parentPath = []) => {
      // Compute some final state for the row
      const path = [...parentPath, index]

      row.path = path
      row.depth = depth

      row.isExpanded =
        (row.original && row.original[expandedKey]) || getBy(expanded, path)

      if (paginateSubRows || (!paginateSubRows && row.depth === 0)) {
        expandedRows.push(row)
      }

      if (row.isExpanded && row.subRows && row.subRows.length) {
        row.subRows.forEach((row, i) => handleRow(row, i, depth + 1, path))
      }

      return row
    }

    rows.forEach((row, i) => handleRow(row, i))

    return expandedRows
  }, [debug, rows, expandedKey, expanded, paginateSubRows])

  const expandedDepth = findExpandedDepth(expanded)

  return {
    ...props,
    toggleExpandedByPath,
    expandedDepth,
    rows: expandedRows
  }
}

function findExpandedDepth(obj, depth = 1) {
  return Object.values(obj).reduce((prev, curr) => {
    if (typeof curr === 'object') {
      return Math.max(prev, findExpandedDepth(curr, depth + 1))
    }
    return depth
  }, 0)
}
