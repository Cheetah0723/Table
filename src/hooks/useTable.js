import React from 'react'
import PropTypes from 'prop-types'
//
import {
  applyHooks,
  applyPropHooks,
  mergeProps,
  flexRender,
  decorateColumnTree,
  makeHeaderGroups,
  flattenBy,
  determineHeaderVisibility,
} from '../utils'

const propTypes = {
  // General
  data: PropTypes.array.isRequired,
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  defaultColumn: PropTypes.object,
  getSubRows: PropTypes.func,
  getRowID: PropTypes.func,
  debug: PropTypes.bool,
}

const renderErr =
  'You must specify a valid render component. This could be "column.Cell", "column.Header", "column.Filter", "column.Aggregated" or any other custom renderer component.'

export const defaultState = {}

const defaultInitialState = {}
const defaultColumnInstance = {}
const defaultReducer = (old, newState) => newState
const defaultGetSubRows = (row, index) => row.subRows || []
const defaultGetRowID = (row, index) => index

export const useTable = (props, ...plugins) => {
  // Validate props
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useTable')

  // Destructure props
  let {
    data,
    columns: userColumns,
    initialState = defaultInitialState,
    state: userState,
    defaultColumn = defaultColumnInstance,
    getSubRows = defaultGetSubRows,
    getRowID = defaultGetRowID,
    reducer = defaultReducer,
    debug,
  } = props

  debug = process.env.NODE_ENV === 'production' ? false : debug

  // But use the users table state if provided
  let [originalState, originalSetState] = React.useState({
    ...defaultState,
    ...initialState,
  })

  const state = React.useMemo(() => {
    if (userState) {
      const newState = {
        ...originalState,
      }
      Object.keys(userState).forEach(key => {
        newState[key] = userState[key]
      })
      return newState
    }
    return originalState
  }, [originalState, userState])

  const setState = React.useCallback(
    (updater, type) => {
      return originalSetState(old => {
        const newState = typeof updater === 'function' ? updater(old) : updater
        return reducer(old, newState, type)
      })
    },
    [reducer]
  )

  // The table instance ref
  let instanceRef = React.useRef({})

  Object.assign(instanceRef.current, {
    ...props,
    data, // The raw data
    state,
    setState, // The resolved table state
    plugins, // All resolved plugins
    hooks: {
      columnsBeforeHeaderGroups: [],
      columnsBeforeHeaderGroupsDeps: [],
      useBeforeDimensions: [],
      useMain: [],
      useRows: [],
      prepareRow: [],
      getTableProps: [],
      getTableBodyProps: [],
      getRowProps: [],
      getHeaderGroupProps: [],
      getHeaderProps: [],
      getCellProps: [],
    },
  })

  // Allow plugins to register hooks
  if (process.env.NODE_ENV === 'development' && debug) console.time('plugins')

  plugins.filter(Boolean).forEach(plugin => {
    plugin(instanceRef.current.hooks)
  })

  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('plugins')

  // Decorate All the columns
  let columns = React.useMemo(
    () => decorateColumnTree(userColumns, defaultColumn),
    [defaultColumn, userColumns]
  )

  // Get the flat list of all columns andllow hooks to decorate
  // those columns (and trigger this memoization via deps)
  let flatColumns = React.useMemo(() => {
    if (process.env.NODE_ENV === 'development' && debug)
      console.time('hooks.columnsBeforeHeaderGroups')

    let newColumns = applyHooks(
      instanceRef.current.hooks.columnsBeforeHeaderGroups,
      flattenBy(columns, 'columns'),
      instanceRef.current
    )

    if (process.env.NODE_ENV === 'development' && debug)
      console.timeEnd('hooks.columnsBeforeHeaderGroups')
    return newColumns
  }, [
    columns,
    debug,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    ...applyHooks(
      instanceRef.current.hooks.columnsBeforeHeaderGroupsDeps,
      [],
      instanceRef.current
    ),
  ])

  // Make the headerGroups
  const headerGroups = React.useMemo(
    () => makeHeaderGroups(flatColumns, defaultColumn),
    [defaultColumn, flatColumns]
  )

  const headers = React.useMemo(() => headerGroups[0].headers, [headerGroups])

  Object.assign(instanceRef.current, {
    columns,
    flatColumns,
    headerGroups,
    headers,
  })

  // Access the row model
  const [rows, flatRows] = React.useMemo(() => {
    if (process.env.NODE_ENV === 'development' && debug)
      console.time('getAccessedRows')

    let flatRows = []

    // Access the row's data
    const accessRow = (originalRow, i, depth = 0, parentPath = []) => {
      // Keep the original reference around
      const original = originalRow

      const rowID = getRowID(originalRow, i)

      // Make the new path for the row
      const path = [...parentPath, rowID]

      const row = {
        original,
        index: i,
        path, // used to create a key for each row even if not nested
        depth,
        cells: [{}], // This is a dummy cell
      }

      flatRows.push(row)

      // Process any subRows
      let subRows = getSubRows(originalRow, i)

      if (subRows) {
        row.subRows = subRows.map((d, i) => accessRow(d, i, depth + 1, path))
      }

      // Override common array functions (and the dummy cell's getCellProps function)
      // to show an error if it is accessed without calling prepareRow
      const unpreparedAccessWarning = () => {
        throw new Error(
          'React-Table: You have not called prepareRow(row) one or more rows you are attempting to render.'
        )
      }
      row.cells.map = unpreparedAccessWarning
      row.cells.filter = unpreparedAccessWarning
      row.cells.forEach = unpreparedAccessWarning
      row.cells[0].getCellProps = unpreparedAccessWarning

      // Create the cells and values
      row.values = {}
      flatColumns.forEach(column => {
        row.values[column.id] = column.accessor
          ? column.accessor(originalRow, i, { subRows, depth, data })
          : undefined
      })

      return row
    }

    // Use the resolved data
    const accessedData = data.map((d, i) => accessRow(d, i))
    if (process.env.NODE_ENV === 'development' && debug)
      console.timeEnd('getAccessedRows')
    return [accessedData, flatRows]
  }, [debug, data, getRowID, getSubRows, flatColumns])

  instanceRef.current.rows = rows
  instanceRef.current.flatRows = flatRows

  // Determine column visibility
  determineHeaderVisibility(instanceRef.current)

  // Provide a flat header list for utilities
  instanceRef.current.flatHeaders = headerGroups.reduce(
    (all, headerGroup) => [...all, ...headerGroup.headers],
    []
  )

  if (process.env.NODE_ENV === 'development' && debug)
    console.time('hooks.useBeforeDimensions')
  instanceRef.current = applyHooks(
    instanceRef.current.hooks.useBeforeDimensions,
    instanceRef.current
  )
  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('hooks.useBeforeDimensions')

  calculateDimensions(instanceRef.current)

  if (process.env.NODE_ENV === 'development' && debug)
    console.time('hooks.useMain')
  instanceRef.current = applyHooks(
    instanceRef.current.hooks.useMain,
    instanceRef.current
  )
  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('hooks.useMain')

  // Each materialized header needs to be assigned a render function and other
  // prop getter properties here.
  instanceRef.current.flatHeaders.forEach(column => {
    // Give columns/headers rendering power
    column.render = (type, userProps = {}) => {
      const Comp = typeof type === 'string' ? column[type] : type

      if (typeof Comp === 'undefined') {
        throw new Error(renderErr)
      }

      return flexRender(Comp, {
        ...instanceRef.current,
        column,
        ...userProps,
      })
    }

    // Give columns/headers a default getHeaderProps
    column.getHeaderProps = props =>
      mergeProps(
        {
          key: ['header', column.id].join('_'),
          colSpan: column.totalVisibleHeaderCount,
        },
        applyPropHooks(
          instanceRef.current.hooks.getHeaderProps,
          column,
          instanceRef.current
        ),
        props
      )
  })

  instanceRef.current.headerGroups.forEach((headerGroup, i) => {
    // Filter out any headers and headerGroups that don't have visible columns
    headerGroup.headers = headerGroup.headers.filter(header => {
      const recurse = headers =>
        headers.filter(header => {
          if (header.headers) {
            return recurse(header.headers)
          }
          return header.isVisible
        }).length
      if (header.headers) {
        return recurse(header.headers)
      }
      return header.isVisible
    })

    // Give headerGroups getRowProps
    if (headerGroup.headers.length) {
      headerGroup.getHeaderGroupProps = (props = {}) =>
        mergeProps(
          {
            key: [`header${i}`].join('_'),
          },
          applyPropHooks(
            instanceRef.current.hooks.getHeaderGroupProps,
            headerGroup,
            instanceRef.current
          ),
          props
        )

      return true
    }
  })

  // Run the rows (this could be a dangerous hook with a ton of data)
  if (process.env.NODE_ENV === 'development' && debug)
    console.time('hooks.useRows')
  instanceRef.current.rows = applyHooks(
    instanceRef.current.hooks.useRows,
    instanceRef.current.rows,
    instanceRef.current
  )
  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('hooks.useRows')

  // The prepareRow function is absolutely necessary and MUST be called on
  // any rows the user wishes to be displayed.

  instanceRef.current.prepareRow = React.useCallback(row => {
    row.getRowProps = props =>
      mergeProps(
        { key: ['row', ...row.path].join('_') },
        applyPropHooks(
          instanceRef.current.hooks.getRowProps,
          row,
          instanceRef.current
        ),
        props
      )

    // Build the visible cells for each row
    row.cells = instanceRef.current.flatColumns
      .filter(d => d.isVisible)
      .map(column => {
        const cell = {
          column,
          row,
          value: row.values[column.id],
        }

        // Give each cell a getCellProps base
        cell.getCellProps = props => {
          const columnPathStr = [...row.path, column.id].join('_')
          return mergeProps(
            {
              key: ['cell', columnPathStr].join('_'),
            },
            applyPropHooks(
              instanceRef.current.hooks.getCellProps,
              cell,
              instanceRef.current
            ),
            props
          )
        }

        // Give each cell a renderer function (supports multiple renderers)
        cell.render = (type, userProps = {}) => {
          const Comp = typeof type === 'string' ? column[type] : type

          if (typeof Comp === 'undefined') {
            throw new Error(renderErr)
          }

          return flexRender(Comp, {
            ...instanceRef.current,
            column,
            row,
            cell,
            ...userProps,
          })
        }

        return cell
      })

    // need to apply any row specific hooks (useExpanded requires this)
    applyHooks(instanceRef.current.hooks.prepareRow, row, instanceRef.current)
  }, [])

  instanceRef.current.getTableProps = userProps =>
    mergeProps(
      applyPropHooks(
        instanceRef.current.hooks.getTableProps,
        instanceRef.current
      ),
      userProps
    )

  instanceRef.current.getTableBodyProps = userProps =>
    mergeProps(
      applyPropHooks(
        instanceRef.current.hooks.getTableBodyProps,
        instanceRef.current
      ),
      userProps
    )

  return instanceRef.current
}

function calculateDimensions(instance) {
  const { headers } = instance

  instance.totalColumnsWidth = calculateHeaderWidths(headers)
}

function calculateHeaderWidths(headers, left = 0) {
  let sumTotalWidth = 0

  headers.forEach(header => {
    let { headers: subHeaders } = header

    header.totalLeft = left

    if (subHeaders && subHeaders.length) {
      header.totalWidth = calculateHeaderWidths(subHeaders, left)
    } else {
      header.totalWidth = Math.min(
        Math.max(header.minWidth, header.width),
        header.maxWidth
      )
    }
    if (header.isVisible) {
      left += header.totalWidth
      sumTotalWidth += header.totalWidth
    }
  })

  return sumTotalWidth
}
