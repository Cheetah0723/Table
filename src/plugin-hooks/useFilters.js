import React from 'react'

import {
  actions,
  getFirstDefined,
  getFilterMethod,
  useMountedLayoutEffect,
  functionalUpdate,
  useGetLatest,
  shouldAutoRemoveFilter,
} from '../utils'
import * as filterTypes from '../filterTypes'

// Actions
actions.resetFilters = 'resetFilters'
actions.setFilter = 'setFilter'
actions.setAllFilters = 'setAllFilters'

export const useFilters = hooks => {
  hooks.stateReducers.push(reducer)
  hooks.useInstance.push(useInstance)
}

useFilters.pluginName = 'useFilters'

function reducer(state, action, previousState, instance) {
  if (action.type === actions.init) {
    return {
      filters: [],
      ...state,
    }
  }

  if (action.type === actions.resetFilters) {
    return {
      ...state,
      filters: instance.initialState.filters || [],
    }
  }

  if (action.type === actions.setFilter) {
    const { columnId, filterValue } = action
    const { flatColumns, userFilterTypes } = instance

    const column = flatColumns.find(d => d.id === columnId)

    if (!column) {
      throw new Error(
        `React-Table: Could not find a column with id: ${columnId}`
      )
    }

    const filterMethod = getFilterMethod(
      column.filter,
      userFilterTypes || {},
      filterTypes
    )

    const previousfilter = state.filters.find(d => d.id === columnId)

    const newFilter = functionalUpdate(
      filterValue,
      previousfilter && previousfilter.value
    )

    //
    if (shouldAutoRemoveFilter(filterMethod.autoRemove, newFilter)) {
      return {
        ...state,
        filters: state.filters.filter(d => d.id !== columnId),
      }
    }

    if (previousfilter) {
      return {
        ...state,
        filters: state.filters.map(d => {
          if (d.id === columnId) {
            return { id: columnId, value: newFilter }
          }
          return d
        }),
      }
    }

    return {
      ...state,
      filters: [...state.filters, { id: columnId, value: newFilter }],
    }
  }

  if (action.type === actions.setAllFilters) {
    const { filters } = action
    const { flatColumns, filterTypes: userFilterTypes } = instance

    return {
      ...state,
      // Filter out undefined values
      filters: functionalUpdate(filters, state.filters).filter(filter => {
        const column = flatColumns.find(d => d.id === filter.id)
        const filterMethod = getFilterMethod(
          column.filter,
          userFilterTypes || {},
          filterTypes
        )

        if (shouldAutoRemoveFilter(filterMethod.autoRemove, filter.value)) {
          return false
        }
        return true
      }),
    }
  }
}

function useInstance(instance) {
  const {
    data,
    rows,
    flatRows,
    flatColumns,
    filterTypes: userFilterTypes,
    manualFilters,
    defaultCanFilter = false,
    disableFilters,
    state: { filters },
    dispatch,
    autoResetFilters = true,
  } = instance

  const setFilter = (columnId, filterValue) => {
    dispatch({ type: actions.setFilter, columnId, filterValue })
  }

  const setAllFilters = filters => {
    dispatch({
      type: actions.setAllFilters,
      filters,
    })
  }

  flatColumns.forEach(column => {
    const {
      id,
      accessor,
      defaultCanFilter: columnDefaultCanFilter,
      disableFilters: columnDisableFilters,
    } = column

    // Determine if a column is filterable
    column.canFilter = accessor
      ? getFirstDefined(
          columnDisableFilters === true ? false : undefined,
          disableFilters === true ? false : undefined,
          true
        )
      : getFirstDefined(columnDefaultCanFilter, defaultCanFilter, false)

    // Provide the column a way of updating the filter value
    column.setFilter = val => setFilter(column.id, val)

    // Provide the current filter value to the column for
    // convenience
    const found = filters.find(d => d.id === id)
    column.filterValue = found && found.value
  })

  const [filteredRows, filteredFlatRows] = React.useMemo(() => {
    if (manualFilters || !filters.length) {
      return [rows, flatRows]
    }

    const filteredFlatRows = []

    // Filters top level and nested rows
    const filterRows = (rows, depth = 0) => {
      let filteredRows = rows

      filteredRows = filters.reduce(
        (filteredSoFar, { id: columnId, value: filterValue }) => {
          // Find the filters column
          const column = flatColumns.find(d => d.id === columnId)

          if (!column) {
            return filteredSoFar
          }

          if (depth === 0) {
            column.preFilteredRows = filteredSoFar
          }

          const filterMethod = getFilterMethod(
            column.filter,
            userFilterTypes || {},
            filterTypes
          )

          if (!filterMethod) {
            console.warn(
              `Could not find a valid 'column.filter' for column with the ID: ${column.id}.`
            )
            return filteredSoFar
          }

          // Pass the rows, id, filterValue and column to the filterMethod
          // to get the filtered rows back
          column.filteredRows = filterMethod(
            filteredSoFar,
            [columnId],
            filterValue
          )

          return column.filteredRows
        },
        rows
      )

      // Apply the filter to any subRows
      // We technically could do this recursively in the above loop,
      // but that would severely hinder the API for the user, since they
      // would be required to do that recursion in some scenarios
      filteredRows = filteredRows.map(row => {
        filteredFlatRows.push(row)
        if (!row.subRows) {
          return row
        }
        return {
          ...row,
          subRows:
            row.subRows && row.subRows.length > 0
              ? filterRows(row.subRows, depth + 1)
              : row.subRows,
        }
      })

      return filteredRows
    }

    return [filterRows(rows), filteredFlatRows]
  }, [manualFilters, filters, rows, flatRows, flatColumns, userFilterTypes])

  React.useMemo(() => {
    // Now that each filtered column has it's partially filtered rows,
    // lets assign the final filtered rows to all of the other columns
    const nonFilteredColumns = flatColumns.filter(
      column => !filters.find(d => d.id === column.id)
    )

    // This essentially enables faceted filter options to be built easily
    // using every column's preFilteredRows value
    nonFilteredColumns.forEach(column => {
      column.preFilteredRows = filteredRows
      column.filteredRows = filteredRows
    })
  }, [filteredRows, filters, flatColumns])

  const getAutoResetFilters = useGetLatest(autoResetFilters)

  useMountedLayoutEffect(() => {
    if (getAutoResetFilters()) {
      dispatch({ type: actions.resetFilters })
    }
  }, [dispatch, manualFilters ? null : data])

  Object.assign(instance, {
    preFilteredRows: rows,
    preFilteredFlatRows: flatRows,
    filteredRows,
    filteredFlatRows,
    rows: filteredRows,
    flatRows: filteredFlatRows,
    setFilter,
    setAllFilters,
  })
}
