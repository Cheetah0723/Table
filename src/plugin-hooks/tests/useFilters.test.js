import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { useTable } from '../../hooks/useTable'
import { useFilters } from '../useFilters'

const data = [
  {
    firstName: 'tanner',
    lastName: 'linsley',
    age: 29,
    visits: 100,
    status: 'In Relationship',
    progress: 50,
  },
  {
    firstName: 'derek',
    lastName: 'perkins',
    age: 40,
    visits: 40,
    status: 'Single',
    progress: 80,
  },
  {
    firstName: 'joe',
    lastName: 'bergevin',
    age: 45,
    visits: 20,
    status: 'Complicated',
    progress: 10,
  },
  {
    firstName: 'jaylen',
    lastName: 'linsley',
    age: 26,
    visits: 99,
    status: 'In Relationship',
    progress: 70,
  },
]

const defaultColumn = {
  Cell: ({ cell: { value }, column: { id } }) => `${id}: ${value}`,
  Filter: ({ column: { filterValue, setFilter } }) => (
    <input
      value={filterValue || ''}
      onChange={e => {
        setFilter(e.target.value || undefined) // Set undefined to remove the filter entirely
      }}
      placeholder="Search..."
    />
  ),
}

function Table({ columns, data }) {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
  } = useTable(
    {
      columns,
      data,
      defaultColumn,
    },
    useFilters
  )

  return (
    <table {...getTableProps()}>
      <thead>
        {headerGroups.map(headerGroup => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map(column => (
              <th {...column.getHeaderProps()}>
                {column.render('Header')}
                {column.canFilter ? column.render('Filter') : null}
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rows.map(
          (row, i) =>
            prepareRow(row) || (
              <tr {...row.getRowProps()}>
                {row.cells.map(cell => (
                  <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                ))}
              </tr>
            )
        )}
      </tbody>
    </table>
  )
}

function App() {
  const columns = React.useMemo(
    () => [
      {
        Header: 'Name',
        columns: [
          {
            Header: 'First Name',
            accessor: 'firstName',
          },
          {
            Header: 'Last Name',
            accessor: 'lastName',
          },
        ],
      },
      {
        Header: 'Info',
        columns: [
          {
            Header: 'Age',
            accessor: 'age',
          },
          {
            Header: 'Visits',
            accessor: 'visits',
          },
          {
            Header: 'Status',
            accessor: 'status',
          },
          {
            Header: 'Profile Progress',
            accessor: 'progress',
          },
        ],
      },
    ],
    []
  )

  return <Table columns={columns} data={data} />
}

test('renders a filterable table', () => {
  const { getAllByPlaceholderText, asFragment } = render(<App />)

  const filterInputs = getAllByPlaceholderText('Search...')

  const beforeFilter = asFragment()

  fireEvent.change(filterInputs[1], { target: { value: 'l' } })

  const afterFilter1 = asFragment()

  fireEvent.change(filterInputs[1], { target: { value: 'er' } })

  const afterFilter2 = asFragment()

  expect(beforeFilter).toMatchDiffSnapshot(afterFilter1)
  expect(afterFilter1).toMatchDiffSnapshot(afterFilter2)
})
