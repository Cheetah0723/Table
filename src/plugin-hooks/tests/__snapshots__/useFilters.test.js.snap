// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a filterable table 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -37,11 +37,11 @@
            colspan="1"
          >
            Last Name
            <input
              placeholder="Search..."
-             value=""
+             value="l"
            />
          </th>
          <th
            class=""
            colspan="1"
@@ -117,78 +117,10 @@
          </td>
          <td
            class=""
          >
            progress: 50
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           firstName: derek
-         </td>
-         <td
-           class=""
-         >
-           lastName: perkins
-         </td>
-         <td
-           class=""
-         >
-           age: 40
-         </td>
-         <td
-           class=""
-         >
-           visits: 40
-         </td>
-         <td
-           class=""
-         >
-           status: Single
-         </td>
-         <td
-           class=""
-         >
-           progress: 80
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           firstName: joe
-         </td>
-         <td
-           class=""
-         >
-           lastName: bergevin
-         </td>
-         <td
-           class=""
-         >
-           age: 45
-         </td>
-         <td
-           class=""
-         >
-           visits: 20
-         </td>
-         <td
-           class=""
-         >
-           status: Complicated
-         </td>
-         <td
-           class=""
-         >
-           progress: 10
          </td>
        </tr>
        <tr
          class=""
        >
`;

exports[`renders a filterable table 2`] = `
Snapshot Diff:
- First value
+ Second value

@@ -37,11 +37,11 @@
            colspan="1"
          >
            Last Name
            <input
              placeholder="Search..."
-             value="l"
+             value="er"
            />
          </th>
          <th
            class=""
            colspan="1"
@@ -91,70 +91,70 @@
          class=""
        >
          <td
            class=""
          >
-           firstName: tanner
+           firstName: derek
          </td>
          <td
            class=""
          >
-           lastName: linsley
+           lastName: perkins
          </td>
          <td
            class=""
          >
-           age: 29
+           age: 40
          </td>
          <td
            class=""
          >
-           visits: 100
+           visits: 40
          </td>
          <td
            class=""
          >
-           status: In Relationship
+           status: Single
          </td>
          <td
            class=""
          >
-           progress: 50
+           progress: 80
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class=""
          >
-           firstName: jaylen
+           firstName: joe
          </td>
          <td
            class=""
          >
-           lastName: linsley
+           lastName: bergevin
          </td>
          <td
            class=""
          >
-           age: 26
+           age: 45
          </td>
          <td
            class=""
          >
-           visits: 99
+           visits: 20
          </td>
          <td
            class=""
          >
-           status: In Relationship
+           status: Complicated
          </td>
          <td
            class=""
          >
-           progress: 70
+           progress: 10
          </td>
        </tr>
      </tbody>
    </table>
  </DocumentFragment>
`;
